File: python/395_LongestSubstringWithAtLeastKRepeatingCharacters.py
class Solution:
    def longestSubstring(self, s: str, k: int) -> int:
        def helper(start, end):
            if end < k: return 0
            
            counts = [0] * 26
            for i in range(start, end): counts[ord(s[i]) - ord('a')] += 1
            for i in range(start, end):
                if counts[ord(s[i]) - ord('a')] < k:
                    mid = i + 1
                    while mid < end and counts[ord(s[mid]) - ord('a')] < k: mid += 1
                    return max(helper(start, i), helper(mid, end))
            
            return end - start
        
        return helper(0, len(s))

File: swift/395_LongestSubstringWithAtLeastKRepeatingCharacters.swift
class Solution {
    func longestSubstring(_ s: String, _ k: Int) -> Int {
        func helper(_ start: String.Index, _ end: String.Index, _ startingPlace: Int, _ endingPlace: Int) -> Int {
            var counts = [Character : Int]()
            var idx = start
            while idx != end {
                if counts[s[idx]] == nil {
                    counts[s[idx]] = 1
                } else {
                    counts[s[idx]]! += 1
                }
                idx = s.index(after: idx)
            }
            
            idx = start
            var i = startingPlace
            while idx != end {
                if counts[s[idx]]! < k {
                    var midIdx = s.index(after: idx)
                    var mid = i + 1
                    while midIdx != end && counts[s[midIdx]]! < k {
                        midIdx = s.index(after: midIdx)
                        mid += 1
                    }
                    return max(helper(start, idx, startingPlace, i), helper(midIdx, end, mid, endingPlace))
                }
                i += 1
                idx = s.index(after: idx)
            }
            return endingPlace - startingPlace
        }
        return helper(s.startIndex, s.endIndex, 0, s.count)
    }
}
